gremlin-groovy/src/main/java/org/apache/tinkerpop/gremlin/groovy/jsr223/ParameterizedGroovyTranslator.java

```diff
@@ -0,0 +1,346 @@
+import org.apache.tinkerpop.gremlin.structure.VertexProperty;
+import org.apache.tinkerpop.gremlin.structure.util.StringFactory;
+import org.apache.tinkerpop.gremlin.util.function.Lambda;
+
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Date;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
+/**
+ * Converts bytecode to a Parameterized Groovy string of Gremlin, include script and parameters
+ * <p>
+ * {@link GroovyTranslator } implementation provide another way to convert bytecode to Groovy string of Gremlin.
+ * <p>
+ * The only difference between ParameterizedGroovyTranslator and GroovyTranslator is that in GroovyTranslator,
+ * the translation process does not extract parameters.
+ * <p>
+ * Example:
+ * <pre>
+ * Input bytecode: g.addV("person").property("name","marko")
+ * GroovyTranslator:
+ *  dsl: g.addV("person").property("name","marko");
+ *  bindings:  empty map
+ * ParameterizedGroovyTranslator:
+ *  dsl: g.addV(_args_0).property(_args_1,_args_2);
+ *  bindings: {"_args_0":"person", "_args_1":"name", "_args_2":"marko"}
+ * </pre>
+ *
+ * @author Marko A. Rodriguez (http://markorodriguez.com)
+ * @author Stephen Mallette (http://stephen.genoprime.com)
+ * @author Stark Arya (sandszhou.zj@alibaba-inc.com)
+ */
+
+public class ParameterizedGroovyTranslator implements Translator.ParameterizedScriptTranslator {
+    private final String traversalSource;
+    private final TypeTranslator typeTranslator;
+
+    private ParameterizedGroovyTranslator(final String traversalSource, TypeTranslator typeTranslator) {
+        this.traversalSource = traversalSource;
+        this.typeTranslator = typeTranslator;
+    }
+
+    public static final ParameterizedGroovyTranslator of(final String traversalSource) {
+        return of(traversalSource,null);
+    }
+
+    public static final ParameterizedGroovyTranslator of(final String traversalSource, final TypeTranslator typeTranslator) {
+        return new ParameterizedGroovyTranslator(traversalSource,
+                Optional.ofNullable(typeTranslator).orElseGet(DefaultTypeTranslator::new));
+    }
+
+    //////
+
+    @Override
+    public ParameterizedScript translate(final Bytecode bytecode) {
+        return (ParameterizedScript) typeTranslator.apply(traversalSource, bytecode);
+    }
+
+    @Override
+    public String getTargetLanguage() {
+        return "gremlin-groovy";
+    }
+
+    @Override
+    public String toString() {
+        return StringFactory.translatorString(this);
+    }
+
+    @Override
+    public String getTraversalSource() {
+        return this.traversalSource;
+    }
+
+    /**
+     * Performs standard type translation for the TinkerPop types to Parameterized Groovy.
+     */
+    public static class DefaultTypeTranslator implements TypeTranslator {
```

Comment:
It bothers me a bit that the `TypeTranslator` is being used for more than just converting types which was its original intent so that providers/users could plugin their own conversions for custom types. If we implement things the way they are now, they not only have to write two separate `TypeTranslators` (if they want to support both forms of translation) but they also have to reason about `Parameters`. I'm not sure what the solution is there, but I think that's the next thing that's concerning me about this change that I don't have a good answer for yet.

Reply:
In most cases, people just need call the following function:
        public static final ParameterizedGroovyTranslator of(final String traversalSource) 
when they need a default parameterized script translation，this cover most cases and easy enough for use.

As you have mentioned, when people want their own ParameterizedScriptTranslator.TypeTranslator Instead of DefaultTypeTranslator, they should known about Parameters，it may exceed TypeTranslator’s original intent boundary。

I think，if we  take paramterized script  as a part script and if people has some demand on paramterized script, also they need their personal paramterized script. it's necessary for them to known about Parameters and  if you worry about this  i think a better way seems like： ParameterizedScript class， we could move：
        public static final class Parameters {
to a separate file like ParameterizedBindings.java for user to directly use.

Reply:
python test fail, and i thinks it's  travis-ci problem